pkg/controller/user-cluster-controller-manager/constraint-template-controller/controller.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2020 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package constrainttemplatecontroller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/open-policy-agent/frameworks/constraint/pkg/apis/templates/v1beta1"
+	"go.uber.org/zap"
+
+	kubermaticapiv1 "k8c.io/kubermatic/v2/pkg/api/v1"
+	v1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// This controller syncs the kubermatic constraint templates to gatekeeper constraint templates on the user cluster.
+	controllerName = "constraint_template_controller"
+)
+
+type reconciler struct {
+	ctx        context.Context
+	log        *zap.SugaredLogger
+	userClient ctrlruntimeclient.Client
+	seedClient ctrlruntimeclient.Client
+	recorder   record.EventRecorder
+}
+
+func Add(ctx context.Context, log *zap.SugaredLogger, userMgr, seedMgr manager.Manager) error {
+	log = log.Named(controllerName)
+
+	r := &reconciler{
+		ctx:        ctx,
+		log:        log,
+		userClient: userMgr.GetClient(),
+		seedClient: seedMgr.GetClient(),
+		recorder:   userMgr.GetEventRecorderFor(controllerName),
+	}
+	c, err := controller.New(controllerName, userMgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %v", err)
+	}
+
+	ctSource := &source.Kind{Type: &v1.ConstraintTemplate{}}
+	if err := ctSource.InjectCache(seedMgr.GetCache()); err != nil {
+		return fmt.Errorf("failed to inject seed cache into watch: %v", err)
+	}
+
+	// Watch for changes to ConstraintTemplates
+	if err = c.Watch(ctSource, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to establish watch for the ConstraintTemplates %v", err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
```

Comment:
I would suggest to move the logic of the reconcile completely in the private method `reconcile`. Starting from line 99 to line 129.  Take a look at the gist https://gist.github.com/moadqassem/4a4cb9d065a4b52942a63b89c47e0126